### ISSUE TITLE
Color schemes: choose scheme based on user preference

### ISSUE THREAD

**Issue by MatushkaRossiya** [squidfunk/mkdocs-material#1266]
## Description

Add support for [prefers-color-scheme](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme) CSS feature.

### Use Cases

MkDocs pages with Material theme will automatically switch between light and dark look based on user's OS preference (Windows 10, macOS 10.14+, Android 10+, iOS 13+, etc), matching system UI, web browser and other apps.

**Comment by squidfunk** [comment]
Thanks for your suggestion.

Material (currently) doesn’t have a dark mode, so this would be the first thing that needs to be implemented. This could be done I a way that it automatically supports the level 5  ‘prefers-color-scheme` and even `light-level` media queries.

However, this would need to be done orthogonally to the current possibility of color customization, so both things can be combined.

I’m happy to discuss concepts and ideas here and help to craft a PR, as I’m not able to do it myself at the moment.

**Comment by Ir1d** [comment]
btw: In [mdui](https://www.mdui.org/docs/#mdui-dialog), when selecting dark-mode, color of the nav-bar is overwritten to black. (it seems that the color customization can be overwriten when set to dark mode. but it is somehow beyond my ability to implement a front-end option to change themes :|

**Comment by squidfunk** [comment]
Partially solved by #1305

**Comment by henrywhitaker3** [comment]
I have added a [second stylesheet](https://github.com/henrywhitaker3/mkdocs-material-dark-theme/blob/master/docs/assets/css/dark-theme-system.css) to my dark theme [repo](https://github.com/henrywhitaker3/mkdocs-material-dark-them) which is just the main dark theme CSS wrapped in the `@media(prefers-color-shceme: dark)` media query. I've only been able to test it on Android 10, but it works on that.

**Comment by squidfunk** [comment]
@henrywhitaker3 probably not necessary to provide a separate stylesheet, as it could be done via `<link ... media="(prefers-color-scheme: dark)" />` which could be configured via a flag in `mkdocs.yml`

**Comment by henrywhitaker3** [comment]
Oh yeah for sure, just didn't/don't have the time to work out how to do that haha

**Comment by Stanzilla** [comment]
I implemented it like this:

```js
!function() {
	const darkMode = document.getElementById('dark-mode-toggle');

	if (darkMode) {
		const isDarkSchemePreferred = () => window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches;

		darkMode.addEventListener('click', function(e) {
			e.preventDefault();
			e.stopImmediatePropagation();

			const isDark = document.documentElement.classList.contains('dark-mode') ? 0 : 1;

			this.href = '?dark=' + (1 - isDark);

			// If prefers-color-scheme is dark, and user disables dark mode, we need to keep the local storage as dark-mode = 0
            document.documentElement.classList.toggle('dark-mode', isDark);
            localStorage.setItem('dark-mode', isDark)
		});

		if (isDarkSchemePreferred() && localStorage.getItem('dark-mode') === -1) {
		    darkMode.click();
		}
	}
}();
```

then after the `<body>` you add 

```js
if (localStorage.getItem("dark-mode") === "1") {
  document.documentElement.classList.add("dark-mode")
}
```

and theming is done via CSS variables.

**Comment by Mattwmaster58** [comment]
@Stanzilla I'm not well versed in the usage of `prefers-color-scheme`, but it looks like your dark mode implementation would by default flash white before going dark. Is it possible to implement the other way around, ie defaults to to `prefer-colors-scheme` and then change to the user specific preference?

**Comment by squidfunk** [comment]
My advise would be to use `extrahead` and this solution: https://github.com/squidfunk/mkdocs-material/issues/1266#issuecomment-541812056

There should be no flashing.

**Comment by Stanzilla** [comment]
@Mattwmaster58 if you add the script directly after the opening body then it won't flash.

**Comment by Mattwmaster58** [comment]
Great! I just wanted to make sure because it's been quite a few dark mode implementations that are *almost* there but flash white men loading the page lol.

**Comment by squidfunk** [comment]
#1639 contains the dark mode implementation, which is still a WIP.

**Comment by squidfunk** [comment]
2491b23a adds experimental support for preference-based color schemes:

``` yaml
theme:
  palette:
    scheme: prefers
```

This will set the theme to `slate` if the `(prefers-color-scheme: dark)` media query evaluates to `true`. The current implementation is experimental - we'll probably need to refactor it when we settle on a solution for general user-preferences as currently discussed in #1732, as there might be some primary/accent color combinations that don't work well with both schemes, but need to be adjusted by themselves. #1732 contains a good but still unfinished proposal.

**Comment by squidfunk** [comment]
[Documentation](https://squidfunk.github.io/mkdocs-material/getting-started/#color-scheme) added in 143cb05e and released as part of 5.3.0.

Note that the value to be set changed from `prefers` to `preference`.
